deepchem/feat/base_classes.py

```diff
@@ -207,6 +210,198 @@ def __call__(self, molecules):
     return self.featurize(molecules)
 
 
+class StructureFeaturizer(Featurizer):
+  """
+  Abstract class for calculating a set of features for an
+  inorganic crystal structure.
+
+  The defining feature of a `StructureFeaturizer` is that it
+  operates on 3D crystal structures with periodic boundary conditions. 
+  Inorganic crystal structures are represented by Pymatgen structure
+  objects. Featurizers for inorganic crystal structures that are subclasses of
+  this class should plan to process input which comes as pymatgen
+  structure objects. 
+
+  This class is abstract and cannot be invoked directly. You'll
+  likely only interact with this class if you're a developer. Child 
+  classes need to implement the _featurize method for calculating 
+  features for a single crystal structure.
+
+  Notes
+  -----
+  Some subclasses of this class will require pymatgen and matminer to be
+  installed.
+
+  """
+
+  def featurize(self, structures: Iterable[JSON],
+                log_every_n: int = 1000) -> np.ndarray:
+    """Calculate features for crystal structures.
+
+    Parameters
+    ----------
+    structures: Iterable[JSON]
+      Iterable sequence of pymatgen structure dictionaries.
+      Json-serializable dictionary representation of pymatgen.core.structure
+      https://pymatgen.org/pymatgen.core.structure.html
+    log_every_n: int, default 1000
+      Logging messages reported every `log_every_n` samples.
+
+    Returns
+    -------
+    features: np.ndarray
+      A numpy array containing a featurized representation of
+      `structures`.
+
+    """
+
+    # Special case handling of single crystal structure
+    if not isinstance(structures, Iterable):
+      structures = [structures]
+    else:
+      # Convert iterables to list
+      structures = list(structures)
+
+    try:
+      from pymatgen import Structure
+    except ModuleNotFoundError:
+      raise ValueError("This class requires pymatgen to be installed.")
+
+    features = []
+    for idx, structure in enumerate(structures):
+      if idx % log_every_n == 0:
+        logger.info("Featurizing datapoint %i" % idx)
+      try:
+        s = Structure.from_dict(structure)
+        features.append(self._featurize(s))
+      except:
+        logger.warning(
+            "Failed to featurize datapoint %i. Appending empty array" % idx)
+        features.append(np.array([]))
+
+    features = np.asarray(features)
+    return features
+
+  def _featurize(self, structure: "pymatgen.Structure"):
```

Comment:
We can't give `pymatgen.Structure` as a return type, because `pymatgen` doesn't get imported into the global scope.